### PR TITLE
Post Tags: Fix bug where no tags are rendered

### PR DESCRIPTION
**Pull request by jeyip**
## Description

This PR addresses a task in #21087 and is meant to add feature parity to FSE blocks. Previously, when the post tags block was added to a view, post tags **were not** being loaded. This PR implements a solution similar to what @Addison-Stavlo did for the post excerpt block in https://github.com/WordPress/gutenberg/pull/23945.

## Important Notes
- From my understanding, tags can only be applied to posts, **not** pages.
  - While viewing a page in the site editor (ex. static home page), and not a post, it seems like there are no tags to display. (Please let me know if this assumption is incorrect)
  - In the situation where post tags blocks are used on a page, I render a warning that informs the user that the post tags block on a page doesn't make sense.
  - An alternative might be to prevent the user from even inserting post tags blocks on pages.
- Adding a post tags block inside of a query loop block will add tags to _all_ posts 

## How has this been tested?

Tested on local docker wp-env environment.

1. Create a post and add tags in the settings sidebar
2. Navigate to the site editor
3. In the template dropdown, navigate to the recently created post
4. Add the post tags block
5. Verify that the correct tags are applied

### Query Loop Index
post-tags-query-loop

### Singular
post-tags-singular

### Static Home Page (Rendering post tags on pages doesn't seem to make sense. Instead I render a warning)
post-tags-static-home-page

## Types of changes

Bug fix (non-breaking change which fixes an issue)

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by ockham** [comment]
> * From my understanding, tags can only be applied to posts, **not** pages.
>   * While viewing a page in the site editor (ex. static home page), and not a post, it seems like there are no tags to display. (Please let me know if this assumption is incorrect)

I think this is accurate -- with a caveat, [plugins can override that behavior](https://www.sitepoint.com/wordpress-pages-use-tags/). But I think it's fine for a first iteration to only look at the post type; maybe a `TODO` comment that mentions that we should check whether the `post_tag` taxonomy is registered for the `page` post type. (We might be able to gather this information from the REST API, but I'm not even sure we really can.)

> * In the situation where post tags blocks are used on a page, I render a warning that informs the user that the post tags block on a page doesn't make sense.
> * An alternative might be to prevent the user from even inserting post tags blocks on pages.

I would find that preferable TBH -- UX-wise, I think it makes sense to hide something from the user that doesn't make sense in the given context, rather than informing them after the fact. I'm doing something similar here for the Post Content Block when used outside of templates and template parts: #24006

**Comment by ockham** [comment]
> (We might be able to gather this information from the REST API, but I'm not even sure we really can.)

Oh, looks like we can: http://localhost:8888/index.php?rest_route=%2Fwp%2Fv2%2Ftaxonomies%2Fpost_tag%2F&context=edit&_locale=user has a `types` field that's set to `[ 'post' ]`. So we would probably need to check that array.

However, I'm not sure we can do that easily during Gutenberg's client-side block registration (i.e. in `__experimentalRegisterExperimentalCoreBlocks`) right now, since network fetches aren't easily available in that function. So let's keep this for a later iteration.

**Comment by ockham** [comment]
> However, I'm not sure we can do that easily during Gutenberg's client-side block registration (i.e. in `__experimentalRegisterExperimentalCoreBlocks`) right now, since network fetches aren't easily available in that function. So let's keep this for a later iteration.

(Alternatively, we might do this server-side, once https://github.com/WordPress/wordpress-develop/pull/419 gets merged.)

**Comment by jeyip** [comment]
> I would find that preferable TBH -- UX-wise, I think it makes sense to hide something from the user that doesn't make sense in the given context, rather than informing them after the fact.

@ockham Your reasoning makes sense. Thanks for the insight into how to make this possible 🙏 

cc @dubielzyk for opinions on how to handle the post tags block

File: packages/block-library/src/post-tags/edit.js
Comment (Addison-Stavlo) [inline packages/block-library/src/post-tags/edit.js:54]: I like this second warning, I want to steal it for the Post Comments block as well.  Although Im not sure if it makes sense in either case given what @ockham was mentioning about plugins overriding the behavior of allowing tags (and possibly comments?) on non-post entities. 🤔

Comment (ockham) [reply packages/block-library/src/post-tags/edit.js:54]: We could maybe rephrase it a bit?

```suggestion
					'Post tags block: Tags are not available for this post type.'
```

**Comment by ockham** [comment]
> > However, I'm not sure we can do that easily during Gutenberg's client-side block registration (i.e. in `__experimentalRegisterExperimentalCoreBlocks`) right now, since network fetches aren't easily available in that function. So let's keep this for a later iteration.
> 
> (Alternatively, we might do this server-side, once [WordPress/wordpress-develop#419](https://github.com/WordPress/wordpress-develop/pull/419) gets merged.)

Exploring an alternative that doesn't need https://github.com/WordPress/wordpress-develop/pull/419 in #24114. If that flies, we can carry the principle over here.

File: packages/block-library/src/post-tags/edit.js
Comment (jeyip) [reply packages/block-library/src/post-tags/edit.js:54]: > We could maybe rephrase it a bit?

Post type makes sense as a developer. Are users generally familiar with the term `post type` as well? If they are, then I think the phrasing is great and I'll modify the text. 👍

If not, then maybe something more user friendly would be "Tags are not available for this view."

**Comment by ockham** [comment]
> Exploring an alternative that doesn't need [WordPress/wordpress-develop#419](https://github.com/WordPress/wordpress-develop/pull/419) in #24114. If that flies, we can carry the principle over here.

Didn't fly :grimacing: Ignore for now :sweat_smile:

File: packages/block-library/src/post-tags/edit.js
Comment (ockham) [reply packages/block-library/src/post-tags/edit.js:54]: Ah yeah, good point. Maybe not '... for this view.' since ideally the copy we'll choose will still ring familiar with people who know what a post type is. '... for this content' seems good. '... for this type of content.' maybe?

**Comment by dubielzyk** [comment]
>cc @dubielzyk for opinions on how to handle the post tags block

I agree with @ockham that we ideally shouldn't allow them to add it to the post. Is it possible to hide it from the block inserter? If we show it, the warning is good, but I'd wanna refine the copy and perhaps add a little warning symbol there. It seems too "not wrong" at the moment.

File: packages/block-library/src/post-tags/edit.js
Comment (Addison-Stavlo) [reply packages/block-library/src/post-tags/edit.js:54]: Since plugins can allow tags for other CPTs maybe its best to just get rid of this second check?  After all, it should fall back to the 'No Tags.' copy in `PostTagsDisplay` when none are found? 🤔

Comment (ockham) [reply packages/block-library/src/post-tags/edit.js:54]: :man_shrugging: idk. For the uninitiated (that most likely don't have a plugin like that installed), it could be helpful to have messaging that makes it clear that they _can't_ have tags (so they don't go looking after just seeing 'No Tags.').

(But this discussion might be moot if we remove the block from the inserter anyway, as just [greenlit](https://github.com/WordPress/gutenberg/pull/24082#issuecomment-662657939) by @dubielzyk?)

**Comment by jeyip** [comment]
> Didn't fly 😬 Ignore for now 😅

A+ for effort 👏

> I agree with @ockham that we ideally shouldn't allow them to add it to the post. Is it possible to hide it from the block inserter?

@ockham It sounds like the way forward might be to prevent `post tags` from being registered in `__experimentalRegisterExperimentalCoreBlocks`. Ideally we would evaluate post tags fetched from the API and conditionally register the `post tags` block like you've done here https://github.com/WordPress/gutenberg/pull/24006/files. Does that sound correct?

> If we show it, the warning is good, but I'd wanna refine the copy and perhaps add a little warning symbol there.

The warning symbol makes sense to me. Would it be okay to add this as an issue to the create design board? It would be nice to have a quick discussion with a broader audience, especially since this would update all instances of warnings in the site editor and post editor.

**Comment by dubielzyk** [comment]
> The warning symbol makes sense to me. Would it be okay to add this as an issue to the create design board and circle back? It seems like an easy change, but it would update all instances of warnings in the site editor and post editor. I think raising this with a broader audience, would be nice to make sure we're not missing anything.

So you'd be using the standard "block render error" that's used for all other blocks as well? If so, then yeah, let's just use that :)

**Comment by jeyip** [comment]
Work is also being done on the post categories block by @Copons. [There are discussions about similar UX decisions (hiding blocks vs. showing warnings)](https://github.com/WordPress/gutenberg/pull/24091#issuecomment-662383655) 

Here is what we know:
- The design preference is to hide blocks from the inserter instead of displaying warnings.
- We should keep the UX between post categories and post tags consistent.
- Hiding blocks is probably outside of the scope of this PR and Jacopo's PR.

How would we feel about creating a separate issue to tackle hiding blocks? We could unblock both PRs and I could follow up on the issue immediately. I'd ensure that hiding both `post tags` and `post categories` is handled in a consistent way. @ockham

**Comment by ockham** [comment]
> Work is also being done on the post categories block by @Copons. [There are discussions about similar UX decisions (hiding blocks vs. showing warnings)](https://github.com/WordPress/gutenberg/pull/24091#issuecomment-662383655)
> 
> Here is what we know:
> 
> * The design preference is to hide blocks from the inserter instead of displaying warnings.
> * We should keep the UX between post categories and post tags consistent.
> * Hiding blocks is probably outside of the scope of this PR and Jacopo's PR.
> 
> How would we feel about creating a separate issue to tackle hiding blocks? We could unblock the current PRs and I could follow up on the new issue immediately. I'd ensure that hiding both `post tags` and `post categories` is handled in a consistent way. @ockham

Sounds good :+1:

File: packages/block-library/src/post-tags/edit.js
Comment (ockham) [inline packages/block-library/src/post-tags/edit.js:57]:
```suggestion
		 * post_tag taxonomy is registered for the current post type.
```

(This block could be used in other Custom Post Types as well.)

File: packages/block-library/src/post-tags/index.php
Comment (ockham) [inline packages/block-library/src/post-tags/index.php:24]: Why do we need the additional `<div />` wrapper? :thinking:

File: packages/block-library/src/post-tags/edit.js
Comment (jeyip) [reply packages/block-library/src/post-tags/edit.js:57]: Good catch! Making the change now.

File: packages/block-library/src/post-tags/index.php
Comment (jeyip) [reply packages/block-library/src/post-tags/index.php:24]: There are margin styles that target the direct children of entry content for the front-end. Without the wrapping div, the direct children for entry content becomes a series of `<a /> tags`, and spacing margins aren't applied.

Screen Shot 2020-07-23 at 1 15 24 PM
Screen Shot 2020-07-23 at 1 15 14 PM

File: packages/block-library/src/post-tags/edit.js
Comment (ockham) [inline packages/block-library/src/post-tags/edit.js:58]: Hmm, I wish `prettier` didn't indent this comment so much :thinking:

File: packages/block-library/src/post-tags/index.php
Comment (ockham) [reply packages/block-library/src/post-tags/index.php:24]: > There are margin styles that target the direct children of entry content for the front-end. Without the wrapping div, the direct children for entry content become a series of `<a /> tags`, and spacing margins aren't applied. Let me know if that doesn't make sense.

Thanks! This does make sense. Sounds like that's done by the theme though? If it is, it could be argued that it's a theme bug (and should be fixed there rather than here) :thinking: 

(My point is: A different theme could have styling that formats direct `<div />` children of entry content weirdly. So there's no guarantee that what we do to a block's rendered output makes it work for all themes; themes also need to play by some rules.)

Comment (ockham) [reply packages/block-library/src/post-tags/index.php:24]: Since I don't want to block this PR any further, how about we roll back these changes for now and carry them over to another issue or PR for discussion? Arguably, it's not directly related to the goal of this PR (fixing bugs where no tags are rendered).

Comment (jeyip) [reply packages/block-library/src/post-tags/index.php:24]: > Sounds like that's done by the theme though?

Ah that's a great point. Digging a little further reveals that the styles originate from the `seedlet` theme.

> If it is, it could be argued that it's a theme bug (and should be fixed there rather than here) 🤔
> A different theme could have styling that formats direct <div /> children of entry content weirdly.

Totally agree. Thanks for the insight @ockham 

> Since I don't want to block this PR any further, how about we roll back these changes for now and carry them over to another issue or PR for discussion?

Sounds good. I'll open an issue on `Automattic/Themes` and roll back the changes now.

File: packages/block-library/src/post-tags/edit.js
Comment (jeyip) [reply packages/block-library/src/post-tags/edit.js:58]: Yeah it felt odd to me as well. I wonder if we can change some rules...lemme check.

File: packages/block-library/src/post-tags/index.php
Comment (jeyip) [reply packages/block-library/src/post-tags/index.php]: https://github.com/Automattic/themes/issues/2288

File: packages/block-library/src/post-tags/edit.js
Comment (jeyip) [reply packages/block-library/src/post-tags/edit.js:58]: After some poking around, it seems like the easiest way to solve this is with `prettier-ignore` comments. https://prettier.io/docs/en/ignore.html They don't seem to work in the block-library though 🤔

Comment (jeyip) [reply packages/block-library/src/post-tags/edit.js:58]: It looks like `prettier-ignore` doesn't work on comment formatting 😅  I think we may be stuck with the extra indentations.

Comment (ockham) [reply packages/block-library/src/post-tags/edit.js:58]: I gave this some more thought. For inspiration, I looked into the anatomy of a pre-FSE theme (Twenty Twenty):

image

A lot of these come with a number of class name conventions (Post Content's `.entry-content` being one of the more popular ones; or those `.meta-...` ones). However, the particular nesting of `<div />`s and attached class names are fairly theme-specific; after all, they depend on the visual structure of the theme.

It's tempting to carry over some of those class names so that block-based themes can be styled. But that might actually be a fallacy; after all, we want themes to be editable in the Site Editor, which means that all styling should be customizable through the block's controls (and block inspector sidebar, and Global Styles), rather than manual CSS editing.

Maybe we should give this block a light block wrapper though, since that seems to be the sort of interface we tend to base some of this customizability on.

File: packages/block-library/src/post-tags/index.php
Comment (Copons) [reply packages/block-library/src/post-tags/index.php]: Sorry to chime in on this so late but, as far as I can see, _most_ (all?) block-first themes handle the positioning of the content at block level to improve support for different block alignment (wide, full).
In other words, this:
```scss
.entry-content {
  margin: 0 auto;
  width: 700px;
}
```
has been replaced by this:
```scss
.entry-content > * {
  margin: 0 auto;
  width: 700px
}
```

For this reason, if we don't wrap a block inside a `div` (or any block element really, see the Paragraph block using just `p`), it won't be positioned appropriately relative to the content.
Furthermore, we do need a wrapper element to apply the block customization classes and styles.

Comment (Copons) [reply packages/block-library/src/post-tags/index.php]: Ah I think @ockham replied [this comment](https://github.com/WordPress/gutenberg/pull/24082#discussion_r460791766) to the wrong thread. 🙂 

> Maybe we should give this block a light block wrapper though, since that seems to be the sort of interface we tend to base some of this customizability on.

Big +1
It's also very easy to implement (once you know what you need to do at least 😄 ).

Comment (ockham) [reply packages/block-library/src/post-tags/index.php]: > Ah I think @ockham replied this comment to the wrong thread. slightly_smiling_face

:man_facepalming: Thanks @Copons, I'll move it over to this thread.

Comment (ockham) [reply packages/block-library/src/post-tags/index.php]: > Ah I think @ockham replied [this comment](https://github.com/WordPress/gutenberg/pull/24082#discussion_r460791766) to the wrong thread. 
> 
> > Maybe we should give this block a light block wrapper though, since that seems to be the sort of interface we tend to base some of this customizability on.
> 
> Big +1
> It's also very easy to implement (once you know what you need to do at least  ).

On second thought, are light block wrappers available in for dynamic blocks (i.e. server-side, PHP rendered)? :thinking:

Comment (Copons) [reply packages/block-library/src/post-tags/index.php]: > On second thought, are light block wrappers available in for dynamic blocks (i.e. server-side, PHP rendered)? 🤔

@ockham As far as I can understand the light block wrapper is supposed to help keep the _editor_ DOM as lean as possible, (among other reasons) in order to look as close as possible like its rendered counterpart, which reduce the differences between editor and front end styles.

So, turning a block into a light block doesn't affect the rendered version (both JS and PHP).
We'll still need to add that `<div>` wrapper in the PHP side of this PR. 🙂

Comment (jeyip) [reply packages/block-library/src/post-tags/index.php]: > Sorry to chime in on this so late

No worries! 🙂 I always learn something new when you comment.

@Copons Would this `Block.div` component surrounding the `post date` block be an example of a [light block wrapper](https://github.com/WordPress/gutenberg/blob/master/packages/block-library/src/post-date/edit.js#L102-L106)? And is there a standardized way to implement them? I had a hard time finding info in MGS.

>So, turning a block into a light block doesn't affect the rendered version (both JS and PHP).

This was a little surprising to me when I first saw it. By rendering information server side and client side, it seems like we're making an unnecessary distinction between blocks created from initial load (PHP) vs. blocks not created from initial load (JS). It doesn't feel like a huge problem, but I'm curious about the reasoning. 

> We'll still need to add that div wrapper in the PHP side of this PR.

If that's the case, then I'm happy to add the div back in 😛

Comment (Copons) [reply packages/block-library/src/post-tags/index.php]: > This was a little surprising to me when I first saw it. By rendering information server side and client side, it seems like we're making an unnecessary distinction between blocks created from initial load (PHP) vs. blocks not created from initial load (JS). It doesn't feel like a huge problem, but I'm curious about the reasoning. (The alternative might be serving all data necessary for initial load from an endpoint. JS / React could fetch and render from there)

In fact, it's kind of the other way around! 😄 
Blocks with a JS render (the `save.js` file) save the HTML output of that function straight in the post content.
Blocks with a PHP render instead only save a "token" of that block (e.g. `<!-- core/foo-bar -->` + attributes) in the post content, and replace it with the PHP output when rendering the post in the front end.

The JS render is "static" (relatively speaking): the content of the block can only be edited in the editor.
The PHP render instead can be used to display dynamic content: the content of the block can contain stuff coming from elsewhere. FSE blocks are a good example: you can add the Post Tags block in the editor, but its content comes from an SQL query.

Ideally, blocks should use the JS render as much as possible, in order to persist their output in the post content as normal HTML, to be easily exported and reused.
At the same time, we can't just persist the tags in the post content, because the block then wouldn't pick up any changes on tags.

So basically the JS render doesn't run in the front end, but only when saving the post.
There may be blocks that need some JS in the front end (say, a carousel block), in which case we would need to enqueue its own "front end JS" file.
I'm not sure if Core Gutenberg has a standard way to implement it, but for example many Jetpack blocks have a `view.js` file with that exact purpose.

Comment (Copons) [reply packages/block-library/src/post-tags/index.php]: > No worries! 🙂 I always learn something new when you comment.

❤️ 

> @Copons Would this `Block.div` component surrounding the `post date` block be an example of a [light block wrapper](https://github.com/WordPress/gutenberg/blob/master/packages/block-library/src/post-date/edit.js#L102-L106)? And is there a standardized way to implement them?

Yes, the Post Date would be a good example.
Basically you need 2 pieces:
- Add support for `lightBlockWrapper `.

https://github.com/WordPress/gutenberg/blob/b67cd9a9ebb621b8711053fcc44f7030c7dfaee4/packages/block-library/src/post-date/block.json#L18

- Output the block as `<Block.div>` (you can choose another tag, e.g. `<Block.p>`, but I don't think it works appropriately with inline elements).

https://github.com/WordPress/gutenberg/blob/b67cd9a9ebb621b8711053fcc44f7030c7dfaee4/packages/block-library/src/post-date/edit.js#L16

https://github.com/WordPress/gutenberg/blob/b67cd9a9ebb621b8711053fcc44f7030c7dfaee4/packages/block-library/src/post-date/edit.js#L102

---

One important (and obvious, in hindsight) thing to keep in mind is that `Block.div` is now the wrapper!

You can keep `BlockControls` and `InspectorControls` outside of it, as they will be automagically rendered where appropriate with the Slot/Fill mechanic.
On the other hand, anything that should appear "inside" the block, must live inside it.

Example: when adding a "post not found" `Warning` in another FSE block, I put it _outside_ the `Block.div` wrapper. What happened was that the warning was rendered in the editor at full width, without any block classes, wrong margins etc.
Here is how I fixed it:
https://github.com/WordPress/gutenberg/blob/b67cd9a9ebb621b8711053fcc44f7030c7dfaee4/packages/block-library/src/post-comments-count/edit.js#L56-L68

Comment (jeyip) [reply packages/block-library/src/post-tags/index.php:24]: - Reintroduced the server-side rendered wrapping div
- Added the light block wrapper to markup in `edit.js`

Comment (Copons) [reply packages/block-library/src/post-tags/index.php:24]: I might be wrong on the technicalities, but I think it might be more clear by peeking in the code view (post editor only), where you can see the raw post content.

When you save a post containing a Paragraph block (which has a "JS render"), the rendered HTML of its [`save` function](https://github.com/WordPress/gutenberg/blob/71bf1b5146d5dd0fc15792c1a98fc19df25b062f/packages/block-library/src/paragraph/save.js) is stored in the post content like this:
```html
<!-- wp:paragraph {"textColor":"primary"} -->
<p class="has-primary-color has-text-color">Lorem ipsum</p>
<!-- /wp:paragraph -->
```
You can view this block in the front end without any additional process: it's just plain old HTML, and will work fine anywhere.
Since the "JS render" only runs on post save, we can't use it to fetch dynamic content.

Conversely, when you use a "dynamic" block such as Post Tags, there's no render happening on post save: the editor simply adds the block's "token" (or comment, or whatever we want to call it 🙂) to the post content:
```html
<!-- wp:post-tags {"foo":"bar"} /-->
```
Once in the front end, the block's PHP render callback will basically take that comment, grabs the attributes object, and replace it with the render callback output.
The downside is that this block will only work on WP.

But... blocks can also have both a JS and PHP render! 😄 
Many Jetpack blocks do it in order to have a fallback saved in the post content in case the block is not available anymore (e.g. if the user uninstalls Jetpack).

Maybe a good way to think about it is that blocks are HTML comments.
They can wrap some HTML content (generated on post save by the `save` function), and/or will be replaced by a PHP render function on the front end.
EASY! 😆

File: packages/block-library/src/post-tags/index.php
Comment (Copons) [inline packages/block-library/src/post-tags/index.php:23]: This is unnecessary, and we should also add the main block's class to the div to be consistent to other blocks (and also to give users a way to select it via CSS).

An example of how we did it for [other blocks](https://github.com/WordPress/gutenberg/blob/71bf1b5146d5dd0fc15792c1a98fc19df25b062f/packages/block-library/src/post-comments/index.php#L34-L39) (we could skip `$classes` but I presume we'll want to add the text align class eventually):

```php
$classes = 'wp-block-post-tags';
$output  = sprintf( '<div class="%1$s">', esc_attr( $classes ) );

// ...or just this for now:
$output = '<div class="wp-block-post-tags">';
```

**Comment by Copons** [comment]
Nitpick aside, this is working well for me.

I don't think we need a wrapper function just to output the `Warning`s — see an [example](https://github.com/WordPress/gutenberg/blob/71bf1b5146d5dd0fc15792c1a98fc19df25b062f/packages/block-library/src/post-comments-form/edit.js) of how we could deal with it in a single function without having the linter complain about conditional hooks. 🙂
At the same time, since we'll eventually need a proper way to check for tags support, I guess we could move on with this PR for now, and iterate on it later (with a solid reusable solution).

**Comment by jeyip** [comment]
> I don't think we need a wrapper function just to output the Warnings — see an example of how we could deal with it in a single function without having the linter complain about conditional hooks.

I initially headed in that direction, but I saw runtime exceptions when postType was not found. I mistakenly thought `useEntityProp` was the cause, which is why I left the warnings separated (early return to avoid the exception).

After double-checking, the issue was actually a missing guard clause, so I'll add that in and remove the wrapper function 🙂

File: packages/block-library/src/post-tags/index.php
Comment (jeyip) [reply packages/block-library/src/post-tags/index.php:23]: Good callout.

> I presume we'll want to add the text align class eventually
👍 I have an in-progress PR where I was planning to handle alignment https://github.com/WordPress/gutenberg/pull/24069.  I'll implement option #1.

**Comment by jeyip** [comment]
> At the same time, since we'll eventually need a proper way to check for tags support, I guess we could move on with this PR for now, and iterate on it later (with a solid reusable solution).

I had planned to [tackle it in a separate issue](https://github.com/WordPress/gutenberg/pull/24082#issuecomment-663142499) along with post categories by leveraging Bernie's work on the [allowed_block_types filter](https://github.com/WordPress/wordpress-develop/pull/419).

File: packages/block-library/src/post-tags/edit.js
Comment (Addison-Stavlo) [inline packages/block-library/src/post-tags/edit.js:40]: I notice this was not added or changed here, but this seems a bit odd.  So if we have a problem finding 1 of the tags in the list, we return no tagLinks at all?  
It seems like this may need to be updated in the future so that it returns all the tags that are successfully found 🤔

**Review by Addison-Stavlo — COMMENTED** [review]
This is looking good to me!

The only thing that really stands out to me is something that was not added or changed in this PR:

File: packages/block-library/src/post-tags/edit.js
Comment (jeyip) [reply packages/block-library/src/post-tags/edit.js:40]: That's a good point 🤔 Since it seems out of scope for this PR, would you feel comfortable if we rolled that UX discussion into https://github.com/WordPress/gutenberg/pull/24069?

File: packages/block-library/src/post-tags/edit.js
Comment (Copons) [inline packages/block-library/src/post-tags/edit.js:20]: I've got an infamous adversion to `let`, with the assumption that usually using `const` can help simplify the code. 😄 

In this case `links` only depends on the existence of `tags` which, if undefined, would throw an error on `map`. It can be solved in several ways.
I often like using Lodash, whose `map` is guarded against undefined:
```js
import { map } from 'lodash';
const links = map( tags, tagId => { /* ... */ } );
// when tags is undefined, returns an empty array
```

Otherwise, we can use the [optional chaining operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining) `?.`:
```js
const links = tags?.map( tagId => { /* ... */ } );
// when tags is undefined, returns undefined
```

Either way, we'd be removing a `let` and an `if`. 🙂

File: packages/block-library/src/post-tags/edit.js
Comment (Copons) [inline packages/block-library/src/post-tags/edit.js:58]: I'd probably move the comment after `} else if ( context.postType !== 'post' ) {`.
It threw me a bit off guard while quickly scanning the code, because I was expecting to see the "target" of the comment _inside_ that block (likely because of the indentation).

**Comment by Copons** [comment]
This is working great for me!
I've left a few very minor suggestions, but this is basically ready!

File: packages/block-library/src/post-tags/edit.js
Comment (jeyip) [reply packages/block-library/src/post-tags/edit.js:20]: @Copons Great suggestion! I'll keep this in my back pocket for next time 🙂. I went in the direction of optional chaining. All things being relatively equal, I tend to prefer native javascript over third-party libraries.

File: packages/block-library/src/post-tags/edit.js
Comment (jeyip) [reply packages/block-library/src/post-tags/edit.js:51]: 👍

Bernie and I had a [discussion](https://github.com/WordPress/gutenberg/pull/24082#discussion_r459720636) about indentation where I tried to remedy this (but couldn't). The formatting is unfortunate, given the [WP style guide's](https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/#comments) rule to keep "targets" below the comment, but this seems like a good place to make an exception.

Making the change now.

**Review by jeyip — COMMENTED** [review]
> This is working great for me!
I've left a few very minor suggestions, but this is basically ready!

@Copons Thanks for reviewing this! I think we're in the home stretch 🎉

File: packages/block-library/src/post-tags/index.php
Comment (vindl) [inline packages/block-library/src/post-tags/index.php:24]: Given that we are declaring `$classes` above, we could just hardcode it within this string and avoid calling `sprintf` and `esc_attr`?

```suggestion
		$output  = '<div class="wp-block-post-tags">';
```

**Review by vindl — APPROVED** [review]
This worked well for me. Nice work @jeyip! :shipit:

File: packages/block-library/src/post-tags/index.php
Comment (jeyip) [reply packages/block-library/src/post-tags/index.php:24]: @Copons and I discussed this earlier, but we were planning to add text alignment to the `post tags` block. https://github.com/WordPress/gutenberg/pull/24082#discussion_r462237222 where `sprintf` would make more sense.

Comment (vindl) [reply packages/block-library/src/post-tags/index.php:24]: Sounds good then. 👍

**Pull request merged**
